### PR TITLE
Working CUDA version

### PR DESCRIPTION
**Pull request by srush**
Introduces a cuda version of the GPTQ kernel. This use triton to write a custom kernel for this code. On my machine it is about 2x faster than CPU :( was expecting it to be faster, but I guess CPU is pretty good. Main features

* Triton kernel is implemented in python and exported as a PTX file
* Called from Rust using fp16 and the CUST frameworks. These are gated in the installation. 
* Had to move things around to make them both work. Main call is with a new feature gpu. 

Right now this copies a bunch of code from the GPTQ datastructure. There is likely a cleaner way to do this. 

This almost killed me. Triton is so buggy and hard to use. 

There is also a test that checks that the CPU and GPU give the same results. However I need to read up on Rust testing to understand how to make this a real test.

### CODE REVIEW

**Comment by srush** [comment]
@rachtsingh whenever you get a chance.

File: src/gptq_cuda.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        let mut x_temp: [[F16; IN]; B] = [[F16(f16::from_f32(0.0)); IN]; B];
+        // Remap the input for ACT_ORDER=True
+        for i in 0..B {
+            for j in 0..IN {
+                x_temp[i][j] = F16(f16::from_f32(x[i][self.g_index[j] as usize]));
+            }
+        }
+
+        let res = || {
+            let xtemp_gpu = x_temp[0].as_dbuf()?;
```
Comment (gaxler) [inline src/gptq_cuda.rs:134]: Each time you call matvec, you do CPU->GPU copy. This operations is very slow, even if you copy only small amount of data, you still incur transfer overheads.

**Comment by gaxler** [comment]
Cool! 
I use a similar approach of PTX export at work, this works pretty well. The issues that your’e getting are Triton related or performance related?

Doing a quick scan of the code I can see you do CPU-> GPU copy before the call and GPU->CPU after the call. This is very slow and blocks your computation.

Best will be to keep everything on the GPU (this means implementing the rest of llama on the GPU). 

Some other ideas  from the top of my head:
- pin the x_temp buffer to [memory](https://docs.rs/cust/0.3.2/cust/memory/index.html#page-locked-host-memory) , this might speed copies, but I would guess the speedup won’t be enough to really utilize the GPU. 
- spread copy overhead across multiple ops - from memory the only thing I can think of is bundling the `Q,K,V` computations as those will use same `x` buffer. IIRC everything else is sequentially dependent.

**Comment by srush** [comment]
Thanks! I'll take a stab at moving the other parts off cpu next.

My bigger worry is that small changes to the Triton kernel (such as using .dot) seem to break the Rust call version, but not the Python call version with the same arguments. This makes me think I am not using the right calling conventions from Cust? 

Is there anything besides grid, shared memory that I need to read from Triton?

File: src/gptq_cuda.rs (hunk 1 above)
Comment (srush) [reply src/gptq_cuda.rs:134]: Yeah, my plan was to first try fusing qkv attention and MLP layer. If I copy DIM values 2 times per layer, GPU should be fast enough. (But we'll see).

**Comment by rachtsingh** [comment]
Unfortunately I’m traveling so can’t do a thorough review or play around with it for a week, so you shouldn’t wait on me to merge.

Some thoughts from reading the code:
1. I haven’t really figured out testing in Rust either but it’s relatively easy to wrap Rust code in Python to write tests, I can give that a crack when I get back.
2. I think compile-time gating is ok for binary size management but makes the code a bit confusing, e.g. `LlamaModel` refers to two different structs depending on the config used, which might make it harder for people to hack on. I think you can achieve the same result by using traits + only having a single impl (via gating).
3. Agree with the above discussion about speed, don’t really have anything new to add - there’s an underlying tradeoff about GPU size users have and inference speed available, but that doesn’t seem to be a bottleneck yet. I’m not super surprised that GPU inference is only 2x faster (so far) given [https://finbarr.ca/how-is-llama-cpp-possible/](https://finbarr.ca/how-is-llama-cpp-possible/), but I’m sure there are speedups (prefill should be very fast on GPU?).

**Comment by gaxler** [comment]
> Thanks! I'll take a stab at moving the other parts off cpu next.
> 
> My bigger worry is that small changes to the Triton kernel (such as using .dot) seem to break the Rust call version, but not the Python call version with the same arguments. This makes me think I am not using the right calling conventions from Cust?
> 
> Is there anything besides grid, shared memory that I need to read from Triton?

What do you mean by break? It crashes at launch or you get unexpected results out?

**Comment by srush** [comment]
@gaxler it runs fine but the result is only partially filled. I've tried a bunch of different grid sizes, but nothing seems to work so far. Made me wonder if I needed to account for stages or something like that.

I'll try to post a small replication of possible.

**Comment by srush** [comment]
@rachtsingh thanks! I agree with all of these. I might push since it still works, and then work on simplicity and speed. 

Likely it will need an entirely different arch to make gpu fast. Although maybe there is a clever way to template the main code path into GPU /CPU .

**Pull request merged**